library/core/src/iter/adapters/flatten.rs

```diff
@@ -426,117 +582,69 @@ where
-            backiter: &'a mut Option<T::IntoIter>,
-            fold: &'a mut impl FnMut(Acc, T::Item) -> R,
-        ) -> impl FnMut(Acc, T) -> R + 'a
-        where
-            T::IntoIter: DoubleEndedIterator,
-        {
-            move |acc, x| {
-                let mut mid = x.into_iter();
-                let r = mid.try_rfold(acc, &mut *fold);
-                *backiter = Some(mid);
-                r
-            }
-        }
-
-        if let Some(ref mut back) = self.backiter {
-            init = back.try_rfold(init, &mut fold)?;
-        }
-        self.backiter = None;
-
-        init = self.iter.try_rfold(init, flatten(&mut self.backiter, &mut fold))?;
-        self.backiter = None;
-
-        if let Some(ref mut front) = self.frontiter {
-            init = front.try_rfold(init, &mut fold)?;
+        fn flatten<U: DoubleEndedIterator, Acc, R: Try<Output = Acc>>(
+            mut fold: impl FnMut(Acc, U::Item) -> R,
+        ) -> impl FnMut(Acc, &mut U) -> R {
+            move |acc, iter| iter.try_rfold(acc, &mut fold)
         }
-        self.frontiter = None;
 
-        try { init }
+        self.iter_try_rfold(init, flatten(fold))
     }
 
     #[inline]
-    fn rfold<Acc, Fold>(self, mut init: Acc, mut fold: Fold) -> Acc
+    fn rfold<Acc, Fold>(self, init: Acc, fold: Fold) -> Acc
     where
         Fold: FnMut(Acc, Self::Item) -> Acc,
     {
         #[inline]
-        fn flatten<T: IntoIterator, Acc>(
-            fold: &mut impl FnMut(Acc, T::Item) -> Acc,
-        ) -> impl FnMut(Acc, T) -> Acc + '_
-        where
-            T::IntoIter: DoubleEndedIterator,
-        {
-            move |acc, x| x.into_iter().rfold(acc, &mut *fold)
-        }
-
-        if let Some(back) = self.backiter {
-            init = back.rfold(init, &mut fold);
-        }
-
-        init = self.iter.rfold(init, flatten(&mut fold));
-
-        if let Some(front) = self.frontiter {
-            init = front.rfold(init, &mut fold);
+        fn flatten<U: DoubleEndedIterator, Acc>(
+            mut fold: impl FnMut(Acc, U::Item) -> Acc,
+        ) -> impl FnMut(Acc, U) -> Acc {
+            move |acc, iter| iter.rfold(acc, &mut fold)
         }
 
-        init
+        self.iter_rfold(init, flatten(fold))
     }
 
     #[inline]
     #[rustc_inherit_overflow_checks]
     fn advance_back_by(&mut self, n: usize) -> Result<(), usize> {
-        let mut rem = n;
-        loop {
-            if let Some(ref mut back) = self.backiter {
-                match back.advance_back_by(rem) {
-                    ret @ Ok(_) => return ret,
-                    Err(advanced) => rem -= advanced,
-                }
-            }
-            match self.iter.next_back() {
-                Some(iterable) => self.backiter = Some(iterable.into_iter()),
-                _ => break,
-            }
+        // Avoid advancing `self.iter` when `n == 0`.
+        if n == 0 {
+            return Ok(());
```

Comment:
This check is actually necessary for correctness now that `advance_by` uses `iter_try_fold`. The `test_flatten_advance_by` test that you added caught this, but here's a simplified example:
```rust
let mut iter = empty::<Range<i32>>().flatten();
iter.advance_by(0).unwrap();
```
If `advance_by` doesn't check for the `0` case then it will call `iter_try_fold`, which will be unable to get hold of a `U` and hence return `try { acc }`, i.e. `ControlFlow::Continue(0)`. The `match` in `advance_by` then translates this into the incorrect `Err(0)`.

Reply:
I think that could be fixed by always calculating the remaining count and then breaking when it reaches zero. That might also be a small indication that returning the remaining amount (as in #92284) is generally better (at least for implementing adapters).

Reply:
Yeah, I could move the branch to the `Continue` branch of the `match`. It's not obvious to me if that's an improvement though — all else being equal, matching the default implementation seemed like the least surprising behavior. But I'm totally willing to move the branch down so `self.iter` gets advanced still in an `advance_by(0)` call, I just don't really know what should decide it, one way or the other.

This is indeed a case where returning the remaining number of elements is easier to implement. Even keeping the `Result<(), usize>` structure intact, returning the remaining number of elements to skip gets rid of both instances of `n -` here. And indeed, returning a `usize` does simplify this case even more. I still do think returning a plain `usize` is a bad idea, though. Perhaps this would overcomplicate it too much, but what about a custom return type that implements `Try` and that can be initialized using a `usize`? In theory, this gives us the best of both worlds.

Reply:
> But I'm totally willing to move the branch down so self.iter gets advanced still in an advance_by(0) call, I just don't really know what should decide it, one way or the other.

Yeah, I think that'll do the job for now (since it is mentioned in the `advance_by` docs) and we can discuss the rest in the tracking issue.